### PR TITLE
Issue 276 - Readonly column navigation, selection, copy

### LINKED ISSUE #276: Copy-paste on possible when the table is editable
As originally mentioned in https://github.com/plotly/dash-table/issues/75#issuecomment-443948092.

> At the moment, copying from browser only seem to work when the table is editable.
> 
> As an example, I tried to copy the content of the cell "New Mexico" on the simple table at https://dash.plot.ly/datatable/, and nothing gets copied in the clipboard.

### PR DESCRIPTION
**Pull request by Marc-Andre-Rivet**
Fixes #276.

- Fix navigation, selection and copy functionality on readonly columns
- Add a new standalone mode (ReadOnly), test against it and modify/add tests to cover the readonly case specifically

@cldougl Not sure it's absolutely necessary for you to get involved -- the tests should cover this pretty well but maybe you'll have some additional insights / ideas.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "dash-table",
-  "version": "3.1.10",
+  "version": "3.1.11",
```
Comment (Marc-Andre-Rivet) [inline package.json]: bump version

File: demo/AppMode.ts [hunk 2]
```diff
@@ -3,13 +3,24 @@ import * as R from 'ramda';
...
     FixedVirtualized = 'fixed,virtualized',
+    ReadOnly = 'readonly',
     Virtualized = 'virtualized'
 }
 
+export const ReadWriteModes = [
+    AppMode.Default,
+    AppMode.FixedVirtualized,
+    AppMode.Virtualized
+];
```
Comment (Marc-Andre-Rivet) [inline demo/AppMode.ts]: Distinguish read/write capable modes from the read only one(s)

File: demo/AppMode.ts [hunk 3]
```diff
@@ -58,12 +72,24 @@ function getDefaultState() {
             merge_duplicate_headers: false,
             row_deletable: true,
             row_selectable: 'single',
-            content_style: 'fit',
+            content_style: ContentStyle.Fit,
             pagination_mode: 'fe'
-        })
+        }) as Partial<PropsWithDefaults>
```
Comment (Marc-Andre-Rivet) [inline demo/AppMode.ts]: Some typing to allow better customization below

File: demo/AppMode.ts [hunk 4]
```diff
@@ -58,12 +72,24 @@ function getDefaultState() {
...
 }
 
+function getReadonlyState() {
+    const state = getDefaultState();
+    state.tableProps.editable = false;
+    state.tableProps.row_deletable = false;
+
+    R.forEach(column => {
+        column.editable = false;
+    }, state.tableProps.columns || []);
```
Comment (Marc-Andre-Rivet) [inline demo/AppMode.ts]: Make the whole table not editable (columns and rows)

File: src/dash-table/handlers/cellEvents.ts [hunk 5]
```diff
@@ -8,16 +8,11 @@ function isCellSelected(selectedCells: SelectedCells, idx: number, i: number) {
...
-        editable,
         selected_cells,
         setProps,
         virtualized
     } = propsFn();
 
-    if (!editable) {
-        return;
-    }
-
```
Comment (Marc-Andre-Rivet) [inline src/dash-table/handlers/cellEvents.ts]: Well.. do handle the click when not editable... 🐳

File: tests/cypress/tests/standalone/delete_row_test.ts [hunk 6]
```diff
@@ -1,8 +1,8 @@
 import DashTable from 'cypress/DashTable';
 
-import { AppMode } from 'demo/AppMode';
+import { ReadWriteModes } from 'demo/AppMode';
 
-Object.values(AppMode).forEach(mode => {
+Object.values(ReadWriteModes).forEach(mode => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/delete_row_test.ts]: The readonly case does not allow deleting rows. Use the readwrite modes instead.

File: tests/cypress/tests/standalone/edit_cell_test.ts [hunk 7]
```diff
@@ -2,9 +2,9 @@ import DashTable from 'cypress/DashTable';
 import DOM from 'cypress/DOM';
 import Key from 'cypress/Key';
 
-import { AppMode } from 'demo/AppMode';
+import { ReadWriteModes } from 'demo/AppMode';
 
-Object.values(AppMode).forEach(mode => {
+Object.values(ReadWriteModes).forEach(mode => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/edit_cell_test.ts]: Same

File: tests/cypress/tests/standalone/navigation_test.ts [hunk 8]
```diff
@@ -28,90 +28,6 @@ Object.values(AppMode).forEach(mode => {
                 DashTable.getCell(3, 1).click();
             });
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/navigation_test.ts]: navigation_test now covers the cases that are applicable to both readonly and readwrite flavors of the table

File: tests/cypress/tests/standalone/readonly_sorting_test.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import { AppMode } from 'demo/AppMode';
+
+Object.values([AppMode.ReadOnly]).forEach(mode => {
+    describe(`sort (readonly), mode=${mode}`, () => {
+        beforeEach(() => {
+            cy.visit(`http://localhost:8080?mode=${mode}`);
+            DashTable.toggleScroll(false);
+        });
+
+        it('can sort', () => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/readonly_sorting_test.ts]: Stuck in implementation details -- the table optimizes the dropdowns into labels when non-editable. The tests needs to be aware of that as the holder of the text value is not the same.

File: tests/cypress/tests/standalone/readwrite_navigation_test.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
+import DashTable from 'cypress/DashTable';
+import DOM from 'cypress/DOM';
+import Key from 'cypress/Key';
+
+import { ReadWriteModes } from 'demo/AppMode';
+
+Object.values(ReadWriteModes).forEach(mode => {
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/readwrite_navigation_test.ts]: Navigation tests that involve the input element -- testing caret, navigation after/while editing, etc.

File: tests/cypress/tests/standalone/readonly_navigation_test.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            describe('into a dropdown cell', () => {
+                beforeEach(() => {
+                    DashTable.getCell(3, 5).click();
+                });
+
+                it('can move', () => {
+                    DOM.focused.type(Key.ArrowRight);
+
+                    DashTable.getCell(3, 6).should('have.class', 'focused');
+                    DOM.focused.type(Key.ArrowLeft, { force: true });
```
Comment (Marc-Andre-Rivet) [inline tests/cypress/tests/standalone/readonly_navigation_test.ts]: Specifically this line https://github.com/plotly/dash-table/pull/293/files#diff-63e03da9cceb32b9aaddc0be5ea11e01R98 is gone

**Comment by cldougl** [comment]
add example app with both editable and not editable cols (and test)

**Pull request merged**